### PR TITLE
Add optional claim limit to FreshmintClaimSale

### PR DESCRIPTION
**Pull request by psiemens**
This PR adds an optional `claimLimit` parameter to the `Sale` object.

### CODE REVIEW

File: cadence/freshmint-claim-sale/FreshmintClaimSale.cdc [hunk 1]
```diff
@@ -290,6 +342,15 @@ pub contract FreshmintClaimSale {
 
             self.checkAllowlist(address: address)
 
+            let claims = self.claims[address] ?? 0
+
+            // Enforce the claim limit if it is set
+            if let claimLimit = self.claimLimit {
+                assert(claims < claimLimit, message: "reached claim limit")
```
Comment (fkenji) [inline cadence/freshmint-claim-sale/FreshmintClaimSale.cdc:349]: Minor / nitpic, could make this assertion after line 352, when you increment the claim count per address. But I think either way works 👍

File: cadence/freshmint-claim-sale/FreshmintClaimSale.cdc [hunk 2]
```diff
@@ -260,14 +310,16 @@ pub contract FreshmintClaimSale {
...
+                    ?? panic("failed to borrow allowlist")
 
                 if let claims = allowlist.getClaims(address: address) {
                     if claims == 0 {
                         panic("address has already consumed all claims")
                     }
 
-                    allowlist.consumeClaim(address: address)
+                    // Reduce the claim count by one
+                    allowlist.setClaims(address: address, claims: claims - 1)
```
Comment (fkenji) [inline cadence/freshmint-claim-sale/FreshmintClaimSale.cdc:322]: Kinda curious on why the function was removed 👀

File: cadence/freshmint-claim-sale/FreshmintClaimSale.cdc [hunk 3]
```diff
@@ -330,7 +391,7 @@ pub contract FreshmintClaimSale {
         /// The integer value is the number of NFTs an  
         /// address is entitled to claim.
         ///
-        access(self) let claimsByAddress: {Address: UInt}
+        pub let claimsByAddress: {Address: UInt}
```
Comment (fkenji) [inline cadence/freshmint-claim-sale/FreshmintClaimSale.cdc:394]: 🤔  Wouldn't it be safer if `claimsByAddress`  remains private? But then again I think everyone that has access to it so far is in a  controlled environment

File: cadence/freshmint-claim-sale/FreshmintClaimSale.cdc [hunk 4]
```diff
@@ -392,6 +437,7 @@ pub contract FreshmintClaimSale {
             receiverPath: receiverPath,
             paymentReceiver: paymentReceiver,
             price: price,
+            claimLimit: claimLimit,
```
Comment (fkenji) [inline cadence/freshmint-claim-sale/FreshmintClaimSale.cdc:440]: Hmm question, could someone keep a reference to allowList and change it at anytime? Do we need to make a copy of it somehow?

**Review by fkenji — APPROVED** [review]
Nice! 👍

File: cadence/freshmint-claim-sale/FreshmintClaimSale.cdc (hunk 2 above)
Comment (psiemens) [reply cadence/freshmint-claim-sale/FreshmintClaimSale.cdc:322]: I realized that `consumeClaim` was redundant. It reads from the `claimsByAddress` dictionary to get the claim count, but we already do that in this function. Removing it means we only read from the dictionary once per claim, which is more performant.

File: cadence/freshmint-claim-sale/FreshmintClaimSale.cdc (hunk 3 above)
Comment (psiemens) [reply cadence/freshmint-claim-sale/FreshmintClaimSale.cdc:394]: Yeah, I'll keep it as private. It shouldn't matter but I also can't think of a good reason to make it public right now (I'm not sure why I changed it).

File: cadence/freshmint-claim-sale/FreshmintClaimSale.cdc (hunk 4 above)
Comment (psiemens) [reply cadence/freshmint-claim-sale/FreshmintClaimSale.cdc:440]: Yeah, that's the intended usage. `allowlist` is a capability that points to the actual allowlist resource, which lives elsewhere in storage. 

This allows a user to create the allowlist before creating the sale. It also lets them use the same allowlist across multiple sales. We assume that the owner of the sale is also the owner of the allowlist and therefore the capability should be private (only accessible by the owner).

**Pull request merged**